components/lcd/esp_lcd_ssd1681/examples/epaper_lvgl_demo/main/main.c

```diff
@@ -0,0 +1,265 @@
+#include "freertos/task.h"
+#include "esp_timer.h"
+#include "esp_lcd_panel_io.h"
+#include "esp_lcd_panel_vendor.h"
+#include "esp_lcd_panel_ops.h"
+#include "driver/gpio.h"
+#include "driver/spi_master.h"
+#include "esp_err.h"
+#include "esp_log.h"
+#include "lvgl.h"
+
+#include "ssd1681_waveshare_1in54_lut.h"
+
+static const char *TAG = "example";
+
+// Using SPI2 in the example
+#define LCD_HOST  SPI2_HOST
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+//////////////////// Please update the following configuration according to your LCD spec //////////////////////////////
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+#define EXAMPLE_LCD_PIXEL_CLOCK_HZ     (20 * 1000 * 1000)
+#define EXAMPLE_PIN_NUM_SCLK           6
+#define EXAMPLE_PIN_NUM_MOSI           7
+#define EXAMPLE_PIN_NUM_MISO           (-1)   // Unused
+#define EXAMPLE_PIN_NUM_EPD_DC         9
+#define EXAMPLE_PIN_NUM_EPD_RST        4
+#define EXAMPLE_PIN_NUM_EPD_CS         10
+#define EXAMPLE_PIN_NUM_EPD_BUSY       18
+
+// The pixel number in horizontal and vertical
+#define EXAMPLE_LCD_H_RES              200
+#define EXAMPLE_LCD_V_RES              200
+
+// Bit number used to represent command and parameter
+#define EXAMPLE_LCD_CMD_BITS           8
+#define EXAMPLE_LCD_PARAM_BITS         8
+
+#define EXAMPLE_LVGL_TICK_PERIOD_MS    2
+
+
+static SemaphoreHandle_t panel_refreshing_sem = NULL;
+
+extern void example_lvgl_demo_ui(lv_disp_t *disp);
+
+IRAM_ATTR bool epaper_flush_ready_callback(const esp_lcd_panel_handle_t handle, const void *edata, void *user_data)
+{
+    lv_disp_drv_t *disp_driver = (lv_disp_drv_t *) user_data;
+    lv_disp_flush_ready(disp_driver);
+    BaseType_t xHigherPriorityTaskWoken = pdFALSE;
+    xSemaphoreGiveFromISR(panel_refreshing_sem, &xHigherPriorityTaskWoken);
+    if (xHigherPriorityTaskWoken == pdTRUE) {
+        return true;
+    }
+    return false;
+}
+
+static void example_lvgl_flush_cb(lv_disp_drv_t *drv, const lv_area_t *area, lv_color_t *color_map)
+{
+    esp_lcd_panel_handle_t panel_handle = (esp_lcd_panel_handle_t) drv->user_data;
+    int offsetx1 = area->x1;
+    int offsetx2 = area->x2;
+    int offsety1 = area->y1;
+    int offsety2 = area->y2;
+    // Used to vertical traverse lvgl framebuffer
+    // int len_x = abs(offsetx1 - offsetx2) + 1;
+    // int len_y = abs(offsety1 - offsety2) + 1;
+    // --- Convert buffer from color to monochrome bitmap
+    int len_bits = (abs(offsetx1 - offsetx2) + 1) * (abs(offsety1 - offsety2) + 1);
+    static uint8_t *converted_buffer_black;
+    static uint8_t *converted_buffer_red;
+    static int flag;
+    if (!flag) {
+        converted_buffer_black = heap_caps_malloc(EXAMPLE_LCD_H_RES * EXAMPLE_LCD_V_RES / 8, MALLOC_CAP_DMA);
+        converted_buffer_red = heap_caps_malloc(EXAMPLE_LCD_H_RES * EXAMPLE_LCD_V_RES / 8, MALLOC_CAP_DMA);
```

Comment:
What about create these buffers in main initialization?

Reply:
Good idea.